### DISCUSSION TITLE
Starcoin Move 合约标准库升级投票指南｜Starcoin Move stdlib upgrade Guide

### DISCUSSION THREAD

**Discussion by jolestar** [starcoinorg/starcoin#2578] [DaoProposal]
# Starcoin Move 合约标准库版升级投票指南

本次 Move 合约标准库（stdlib）升级是 Starcoin 主网上线以来第一次升级，主要包含以下特性：

1. 从国库提款的时候增加额度限制，最大数额不能超过投票通过阈值（当前流通量的 4%）。(https://github.com/starcoinorg/starcoin/pull/2566)
2. 实现了新的链上认证策略，简化初始化链上账号的复杂度。当前，如果账号在链上不存在时，需要同时传递该账号的 Authentication Key 进行账号初始化，使用复杂度较高。本次更新中，初始化账号的时候只需要 address，Authentication Key 直接使用 address。 address 的长度是 128 位，Authentication Key 的长度是 256 位，安全性虽然比 Authentication Key 有所降低，但 128 位的碰撞概率也足够低，可以保证安全。当新创建账号第一次发起交易的时候，会自动根据交易的 public key 生成 256 位的  Authentication Key 并更新到链上。(https://github.com/starcoinorg/starcoin/pull/2562)

## 升级提案信息

1. 升级包二进制：https://github.com/starcoinorg/starcoin/tree/master/vm/stdlib/compiled/5/4-5
2. Package 哈希：0x20d79acaca9c50d4cbf51a992e5de658dcecf39c1573244f7d850b4b47af56d360
3. 提案 id：0
4. 提案交易: https://explorer.starcoin.org/main/transactions/detail/0x0237dba2eb4ea5971f7fb53693acb91f6879bfa12db3c15f1a6281e1661d9ee0
5. 提案发起账号：0xb2aa52f94db4516c5beecef363af850a

## 升级流程

1.  从今天发起升级提案开始，进行为期 7 天的投票。
2. 7 天后如果投票通过，则进入公示期，公示期 1 天。如果未通过，则提案失败。
3. 公示期过后，任何人都可以发起升级合约交易（升级交易的 Package hash 必须等于提案中的 Package hash），将标准库的 v5 版本写入到链上。该升级交易执行后，新特性激活。由于新特性包含了硬分叉特性，届时尚未升级到 starcoin v1.2.0 版本的节点可能会和主网分叉。

### 关于 Starcoin 治理的更多信息请参看：

* [Starcoin 技术白皮书中的治理部分](https://developer.starcoin.org/zh/sips/sip-2/)
* [链上治理的生命周期](https://developer.starcoin.org/zh/key_concepts/dao_governance/)
* [通过治理机制修改 DAO 的设置](https://developer.starcoin.org/zh/cli/modify_dao_config/)

当前，投票治理的 DApp 尚在开发中，需要通过 CLI 命令进行投票。

注意：投票时，会将当前账户的 STC 抵押到合约中，直到投票结束，当前投票周期为 7 天。期间主动退款则视为放弃投票。** 为了鼓励参与链上治理，未来会推出链上治理的激励措施。本次参与投票的用户会根据激励策略进行补偿。**

### 准备
1. 下载最新的 starcoin 版本，并同步主网的区块，进入 starcoin 控制台。
2. 将自己的账户导入到节点内置的钱包中，或者从交易所提币到节点钱包的地址中。

### 操作命令

1. 查看投票状态

```
dev call --function 0x1::Dao::proposal_state -t 0x1::STC::STC -t 0x1::UpgradeModuleDaoProposal::UpgradeModuleV2 --arg 0xb2aa52f94db4516c5beecef363af850a --arg 0
```
第一个参数是提案发起人的地址，第二个参数是提案 id，返回的状态为 1 表示等待状态， 2 表示正在投票中，4 表示投票通过。

2. 执行投票

通过一下命令进行投票：

```
account execute-function -s your_address --function 0x1::DaoVoteScripts::cast_vote -t 0x1::STC::STC -t 0x1::UpgradeModuleDaoProposal::Up
gradeModuleV2 --arg 0xb2aa52f94db4516c5beecef363af850a --arg 0 --arg true --arg 1000000000u128 -b
```

第一个参数是提案发起人的地址，第二个参数是提案 id，第三个参数通过 true/false 表示同意和反对，第四个参数表示用来投票的 STC 数量，注意单位是 nanoSTC，有9位是小数点后的单位，上面的例子是 1 STC。

注意：替换 your_address 为你自己的地址，不要用全部的 STC，要留够 gas 费。

3. 放弃投票

如果想放弃投票，取回抵押的 STC，可以用以下命令：

```
account execute-function -s your_address --function 0x1::DaoVoteScripts::revoke_vote -t 0x1::STC::STC -t 0x1::UpgradeModuleDaoProposal::UpgradeModuleV2 --arg 0xb2aa52f94db4516c5beecef363af850a --arg 0 -b
```

4. 投票结束后取回抵押的 STC

```
account execute-function -s your_address --function 0x1::DaoVoteScripts::unstake_vote -t 0x1::STC::STC -t 0x1::UpgradeModuleDaoProposal::UpgradeModuleV2 --arg 0xb2aa52f94db4516c5beecef363af850a --arg 0 -b
```

5. 查看投票结果

```
dev call --function 0x1::Dao::proposal_info -t 0x1::STC::STC -t 0x1::UpgradeModuleDaoProposal::UpgradeModuleV2 --arg 0xb2aa52f94db4516c5beecef363af850a

{
  "ok": [
    {
      "U64": "0"
    },
    {
      "U64": "1623139659856"
    },
    {
      "U64": "1623744459856"
    },
    {
      "U128": "1000000000"
    },
    {
      "U128": "0"
    }
  ]
}
```

返回值的第一个数表示提案 id，第二个数是开始时间，第三个数是结束时间，第四个数是投票同意的票数，第五个数是投票反对的票数。

--------------
## English Guide

Voting for the Starcoin Move stdlib v5 upgrade is open. This upgrade is the first stdlib upgrade since the Starcoin mainnet lanuched and includes the following features.

1. Add limit on treasury withdrawals proposol amount, the maximum amount of which cannot exceed the voted threshold (4% of the current liquidity). (https://github.com/starcoinorg/starcoin/pull/2566)
2. Implemented a new on-chain authentication strategy to simplify the complexity of initializing on-chain accounts. (https://github.com/starcoinorg/starcoin/pull/2562)

## Upgrade proposal information

1. Upgrade package binary: https://github.com/starcoinorg/starcoin/tree/master/vm/stdlib/compiled/5/4-5
2. Package hash: 0x20d79acaca9c50d4cbf51a992e5de658dcecf39c1573244f7d850b4b47af56d360
3. Proposal id: 0
4. Proposal transaction: https://explorer.starcoin.org/main/transactions/detail/0x0237dba2eb4ea5971f7fb53693acb91f6879bfa12db3c15f1a6281e1661d9ee0
5. Proposer address: 0xb2aa52f94db4516c5beecef363af850a

Coin holders are invited to vote on the chain. When voting, the STC of the current account will be stake to the contract until the end of the voting period, which is currently 7 days.  ** To encourage participation in on-chain governance, incentives for on-chain governance will be introduced in the future. Users who participate in voting this time will be compensated according to the incentive strategy. **

### Preparation

1. Download the latest starcoin version and join the mainnet.
2. Import your account into the node's built-in wallet or withdraw coins from the exchange to the node's wallet address.

### Operation commands

1. Check the voting status

```
dev call --function 0x1::Dao::proposal_state -t 0x1::STC::STC -t 0x1::UpgradeModuleDaoProposal::UpgradeModuleV2 --arg 0xb2aa52f94db4516c5beecef363af850a --arg 0
```
The first argument is the address of the proposer, the second argument is the proposal id, and the returned status is 1 for pending status, 2 for voting, and 4 for voting passed.

2. Execute voting

Voting is performed by the following command.

```
account execute-function -s your_address --function 0x1::DaoVoteScripts::cast_vote -t 0x1::STC::STC -t 0x1:: UpgradeModuleDaoProposal::UpgradeModuleV2 --arg 0xb2aa52f94db4516c5beecef363af850a --arg 0 --arg true --arg 1000000000u128 -b
```

The first argument is the proposer, the second argument is the proposal id, the third argument indicates agreement and disagreement via true/false, and the fourth argument indicates the number of STCs used to vote, note that the unit is nanoSTC with 9 decimal places, the above example is 1 STC.

Note: Replace your_address with your own address, do not use all the STCs, but leave enough for the gas fee.

3. Revoke Voting

If you want to abort the poll and get back the STCs, you can use the following command.

```
account execute-function -s your_address --function 0x1::DaoVoteScripts::revoke_vote -t 0x1::STC::STC -t 0x1:: UpgradeModuleDaoProposal::UpgradeModuleV2 --arg 0xb2aa52f94db4516c5beecef363af850a --arg 0 -b
```

4. retrieve the staked STC after the poll

```
account execute-function -s your_address --function 0x1::DaoVoteScripts::unstake_vote -t 0x1::STC::STC -t 0x1:: UpgradeModuleDaoProposal::UpgradeModuleV2 --arg 0xb2aa52f94db4516c5beecef363af850a --arg 0 -b
```

5. View poll results

```
dev call --function 0x1::Dao::proposal_info -t 0x1::STC::STC -t 0x1::UpgradeModuleDaoProposal::UpgradeModuleV2 --arg 0xb2aa52f94db4516c5beecef363af850a

{
  "ok": [
    {
      "U64": "0"
    },
    {
      "U64": "1623139659856"
    },
    {
      "U64": "1623744459856"
    },
    {
      "U128": "1000000000"
    },
    {
      "U128": "0"
    }
  ]
}
```

The first number of the return value represents the proposal id, the second number is the start time, the third number is the end time, the fourth number is the number of votes for and the fifth number is the number of votes against.

**Comment by wk3368** [comment]
> 本次更新中，初始化账号的时候只需要 address，Authentication Key 直接使用 address。

在客户端，发送stc的时候，有下面两个地方需要receiver的authKey

1. dryRun的时候，

```
const txnRequest = {
        chain_id: chainId,
        gas_unit_price: 1,
        sender: fromAccount,
        sender_public_key: senderPublicKeyHex,
        sequence_number: senderSequenceNumber,
        max_gas_amount: 10000000,
        script: {
          code: '0x1::TransferScripts::peer_to_peer',
          type_args: ['0x1::STC::STC'],
          args: [addHexPrefix(receiverAddress), `x"${receiverAuthKey}"`, sendAmountString],
        },
      }
```

2. sign的时候，RawUserTransaction->TransactionPayload->ScriptFunction->args: 
[addHexPrefix(receiverAddress), `x"${receiverAuthKey}"`, sendAmountString]

你这次的修改可不可以这样理解： 

上面两种情况的args数组的第二个参数receiverAuthKey，
1. 如果receiver 在链上存在，receiverAuthKey可以为空
2. 如果receiver 在链上不存在，receiverAuthkey 可以为空， 也可以不为空(向后兼容)。

反正从现在开始， 客户端不用先去检查receiver是否在链上存在了。

**Reply by jolestar** [reply]
对，现在可以把检查链上是否存在的逻辑给去掉

1. stdlib 升级完成之前，如果账号链上不存在，auth key 为空，则交易会失败。
2. stdlib 升级完成之后，如果账号链上不存在，auth key 为空，则交易会成功。

**Reply by wk3368** [reply]
ok，下下个版本就会去掉！

**Comment by jolestar** [comment]
升级完成

https://explorer.starcoin.org/main/transactions/detail/0x7d4275ca7d444c57ffdccfd0baf6bb5f5f5d2abca90d3763d6b2a93dc046b1e3
